Chaincase.UI/ViewModels/CoinJoinViewModel.cs

```diff
@@ -0,0 +1,391 @@
+                }
+                else
+                {
+                    RoundPhaseState = new RoundPhaseState(RoundPhaseState.Phase, true);
+                }
+                this.RaisePropertyChanged(nameof(RoundPhaseState));
+                this.RaisePropertyChanged(nameof(RoundTimesout));
+                PeersRegistered = mostAdvancedRound.State.RegisteredPeerCount;
+                PeersNeeded = mostAdvancedRound.State.RequiredPeerCount;
+            }
+        }
+
+        private void UpdateRequiredBtcLabel(ClientRound registrableRound)
+        {
+            if (Global.WalletManager is null)
+            {
+                return; // Otherwise NullReferenceException at shutdown.
+            }
+
+            if (registrableRound == default)
+            {
+                if (RequiredBTC == default)
+                {
+                    RequiredBTC = Money.Zero;
+                }
+            }
+            else
+            {
+                var coins = Global.Wallet.Coins;
+                var queued = coins.CoinJoinInProcess();
+                if (queued.Any())
+                {
+                    RequiredBTC = registrableRound.State.CalculateRequiredAmount(Global.Wallet.ChaumianClient.State.GetAllQueuedCoinAmounts().ToArray());
+                }
+                else
+                {
+                    var available = coins.Confirmed().Available();
+                    RequiredBTC = available.Any()
+                        ? registrableRound.State.CalculateRequiredAmount(available.Where(x => x.AnonymitySet < Global.Config.PrivacyLevelStrong).Select(x => x.Amount).ToArray())
+                        : registrableRound.State.CalculateRequiredAmount();
+                }
+            }
+        }
+
+        private void SetBalance()
+        {
+            try
+            {
+				Balance = (
+					Enumerable.Where(Global.Wallet.Coins,
+						c => c.Unspent && !c.SpentAccordingToBackend
+					).Sum(c => (long?)c.Amount) ?? 0
+					).ToString();
+			}
+            catch (Exception error) {
+                Logger.LogError($"CoinJoinViewModel.SetBalance(): Failed to retrieve balance {error} ");
+                Balance = "0";
+            }
+        }
+
+        private void SetUserErrorMessage(string err) {
+            _toastErrorMessage = err;
+            _shouldShowErrorToast = true;
+        }
+
+        private bool IsPasswordValid(string password)
+        {
+            string walletFilePath = Path.Combine(Global.WalletManager.WalletDirectories.WalletsDir, $"{Global.Network}.json");
+            ExtKey keyOnDisk;
+            try
+            {
+                keyOnDisk = KeyManager.FromFile(walletFilePath).GetMasterExtKey(password ?? "");
+            }
+            catch
+            {
+                // bad password
+                return false;
+            }
+            return true;
+        }
+
+        public async void JoinRound(string password) {
```

Comment:
[avoid async void](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#avoid-async-void)

What was wrong with `DoEnqueueAsync` ?

Reply:
I'm not sure. I'll have to revisit that

Reply:
going back to DoEnqueuAsync should fix the ui thread blocking issue too